Editor/Utils/MonoCecilHelper.cs

```diff
@@ -16,5 +17,20 @@ public static IEnumerable<TypeDefinition> AggregateAllTypeDefinitions(IEnumerabl
             }
             return typeDefs;
         }
+        
+        public static bool IsMonoBehaviour(TypeDefinition typeDefinition)
+        {
+            if (typeDefinition.BaseType.FullName.Equals("UnityEngine.MonoBehaviour"))
+                return true;
+
+            try
+            {
+                return IsMonoBehaviour( typeDefinition.BaseType.Resolve());
+            }
+            catch (Exception e)
+            {
+                return false;
```

Comment:
that's right. The problem is that if BaseType is System.Object, it's not null. I can remove the exception handling by checking whether typeDefinition.BaseType.FullName.GetHashCode() corresponds to "System.Object".GetHashCode(). In that case I stop the recursion. That seems to work.

Reply:
looks good to me

Reply:
Actually, the original reason I added the try/catch was this exception:
Mono.Cecil.AssemblyResolutionException : Failed to resolve assembly: 'UnityEngine.CoreModule, Version=0.0.0.0, Culture=neutral, PublicKeyToken=null'

here is another one:
AssemblyResolutionException: Failed to resolve assembly: 'Rewired_Core, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null'
Mono.Cecil.BaseAssemblyResolver.Resolve (Mono.Cecil.AssemblyNameReference name, Mono.Cecil.ReaderParameters parameters)
Mono.Cecil.BaseAssemblyResolver.Resolve (Mono.Cecil.AssemblyNameReference name)
Mono.Cecil.DefaultAssemblyResolver.Resolve (Mono.Cecil.AssemblyNameReference name)
Mono.Cecil.MetadataResolver.Resolve (Mono.Cecil.TypeReference type)
Mono.Cecil.ModuleDefinition.Resolve (Mono.Cecil.TypeReference type)
Mono.Cecil.TypeReference.Resolve ()

Reply:
it looks wrong to me to swallow exceptions. 

questions:

1. Why are we failing to resolve these assemblies? Are you not setting up Cecil's assembly resolver paths? (see https://github.com/jbevain/cecil/blob/master/Mono.Cecil/ModuleDefinition.cs#L29)

2. Even if those exceptions were to be deemed as valid (and expected), why swallow **all** exceptions? It would be much better to catch **AssemblyResolutionException** (anyway, it looks completely wrong to me to catch these exceptions)

Reply:
Also, when checking for a bunch of types from BCL you can use Module.[TypeSystem](https://github.com/jbevain/cecil/blob/master/Mono.Cecil/TypeSystem.cs#L207)

Reply:
> It would be much better to check whether typeDefinition.BaseType != null instead.
fixed: https://github.com/mtrive/ProjectAuditor/commit/c660b8460f65aa94d1a51c63ad7e47f5cd46e281

> Are you not setting up Cecil's assembly resolver paths ?

I was not. Now I am: https://github.com/mtrive/ProjectAuditor/commit/21469792d071e824b26dcd2e67e6ee2fe0967ac4

> why swallow all exceptions? It would be much better to catch AssemblyResolutionException (anyway, it looks completely wrong to me to catch these exceptions)

You are right. So, now that the Resolver fixed the exceptions. Do you think I should still protect from AssemblyResolutionException ?

Reply:
> Are you not setting up Cecil's assembly resolver paths ?

I still think you should not get AssemblyResolutionExceptions. 
anyway, it is much better than before (IMHO) :)

Reply:
The Search Directories for the resolver are setup based on CompilationPipeline.GetPrecompiledAssemblyPaths which is only available as of 2019.x
in 2018 or older, the Resolver is missing some search directories, hence the exception. It does not cause any problem to the user but indeed needs to be fixed.